fs/src/file_info.rs

```diff
@@ -0,0 +1,33 @@
+use std::{fs::File, io, path::PathBuf};
+
+/// Open `File` coupled with its filesystem location and most useful information
+///
+/// The attached context for the `File` is kept minimal to make it easy to construct
+/// without unnecessary kernel queries, but allowing users to:
+/// * associate the file received in callbacks to the request (by its path)
+/// * get the file's most useful metadata information
+#[derive(Debug)]
+pub struct FileInfo {
+    pub file: File,
+    pub path: PathBuf,
+    pub size: u64,
+}
+
+impl FileInfo {
+    /// Create new instance by opening a file from a given `path` and reading its metadata
+    pub fn read_from_path(path: impl Into<PathBuf>) -> io::Result<Self> {
+        let path = path.into();
+        let file = File::open(&path)?;
+        Self::read_from_path_and_file(path, file)
+    }
+
+    /// Create new instance by using already open `file` and only reading its metadata
+    pub fn read_from_path_and_file(path: impl Into<PathBuf>, file: File) -> io::Result<Self> {
+        let size = file.metadata()?.len();
+        Ok(Self {
+            path: path.into(),
+            size,
+            file,
+        })
+    }
```

Comment:
nit: For me, when I see a function that starts with `read`, I first expect it to be related to the [Read trait](https://doc.rust-lang.org/std/io/trait.Read.html). These here subvert that expectation though.

Coupled with how Rust idiomatically calls its constructors `new` or `with`, I think we could rename the fns. Wdyt?


```suggestion
    /// Create new instance by opening a file from a given `path` and reading its metadata
    pub fn new_from_path(path: impl Into<PathBuf>) -> io::Result<Self> {
        let path = path.into();
        let file = File::open(&path)?;
        Self::read_from_path_and_file(path, file)
    }

    /// Create new instance by using already open `file` and only reading its metadata
    pub fn new_from_path_and_file(path: impl Into<PathBuf>, file: File) -> io::Result<Self> {
        let size = file.metadata()?.len();
        Ok(Self {
            path: path.into(),
            size,
            file,
        })
    }
```

Reply:
I agree with the concern - I wanted to highlight that those function will do syscalls (to open the path / read metadata) and just using `new_` in my mind kind of suggest some cheap initialization logic.
But I guess it's also fairly common for `new_*` styles functions to do heavier logic including IO, so I'm fine with renaming it this way - could also consider using `init_from_path` and `init_from_path_and_file`